[clang][Sema] Fix incomplete C mode incompatible ExtInfo/ExtProtoInfo conversion diagnostic

brunodf-gf opened the pull request:
To raise -Wincompatible-function-pointer-type for extended function type information conversions in C mode, checkPointerTypesForAssignment contained an invocation of IsFunctionConversion that is quite dubious: it verifies the conversion from LHS to RHS, while we are converting from RHS to LHS. Presumably, this was done because result and argument types don't have to match exactly in C mode (compatible types is sufficient), while IsFunctionConversion checks that an exact match in function types can be obtained by transforming extended function type information. As a result of the dubious invocation, the diagnostic for incompatible extended function type changes was incomplete and would illogically disappear when other changes to the function type were involved (this was already raised in 2019 but left unsolved). Newer additions of extended function type information such as "cfi_unchecked_callee", ARM CMSE attributes, ARM SME attributes, and function effect attributes, worked around the issue by introducing either holes or additional checks which complicates the logic.

Here, we solve this longstanding issue by using IsFunctionConversion to verify the extended function type information conversion in the correct direction (from RHS to LHS), after first removing other differences in the function type (argument and result type mismatch, prototype vs. no prototype) to allow IsFunctionConversion to obtain an exact match. The complications introduced for checking newer attributes are removed.

Fixes #41465
Fixes #85415

brunodf-gf committed:
[clang][Sema] Fix incomplete C mode incompatible ExtInfo/ExtProtoInfo conversion diagnostic

To raise -Wincompatible-function-pointer-type for extended function type
information conversions in C mode, checkPointerTypesForAssignment
contained an invocation of IsFunctionConversion that is quite dubious:
it verifies the conversion from LHS to RHS, while we are converting from
RHS to LHS. Presumably, this was done because result and argument types
don't have to match exactly in C mode (compatible types is sufficient),
while IsFunctionConversion checks that an exact match in function types
can be obtained by transforming extended function type information. As a
result of the dubious invocation, the diagnostic for incompatible
extended function type changes was incomplete and would illogically
disappear when other changes to the function type were involved (this
was already raised in 2019 but left unsolved). Newer additions of
extended function type information such as "cfi_unchecked_callee" and
function effect attributes, worked around the issue by introducing more
complex checking logic that could open other holes.

Here, we solve this longstanding issue by using IsFunctionConverions to
verify the extended function type information conversion in the correct
direction (from RHS to LHS), after first removing other differences in
the function type (argument and result type mismatch, prototype vs. no
prototype) to allow IsFunctionConversion to obtain an exact match. The
complications for checking "cfi_unchecked_callee" and function effect
attributes are removed.

Fixes #160474

cor3ntin commented:
@Sirraide

Sirraide commented:
Haven’t looked at this in detail yet, but this reminds me of #85415, where we noticed some weirdness around this in sema but we never got around to investigating it.

Sirraide reviewed (commented):
I’m feeling somewhat conflicted about this approach: on the one hand it’s better than what we were doing before, but on the other rebuilding a bunch of types to circumvent our own analysis... a bit weird. I wonder if we could instead extract the parts of `IsFunctionConversion()` that we care about here into a separate function.

brunodf-snps commented:
> Haven’t looked at this in detail yet, but this reminds me of #85415, where we noticed some weirdness around this in sema but we never got around to investigating it.

This is definitely the same root cause.

brunodf-snps commented:
> I’m feeling somewhat conflicted about this approach: on the one hand it’s better than what we were doing before, but on the other rebuilding a bunch of types to circumvent our own analysis... a bit weird. I wonder if we could instead extract the parts of `IsFunctionConversion()` that we care about here into a separate function.

I agree that rebuilding the types is somewhat complex to reuse the logic from `IsFunctionConversion`. But I did not want to duplicate essentially the same logic, and I did not see a way to organize the logic in a shared separate function without (fundamentally) changing the way `IsFunctionConversion` works. Since that would also affect the other callers of this function, I did not want to do this here. So in short, I did not immediately see a better alternative, and it certainly seemed better than the status quo.

Sirraide commented on clang/lib/Sema/SemaExpr.cpp:
```diff
@@ -9187,12 +9187,43 @@ static AssignConvertType checkPointerTypesForAssignment(Sema &S,
+      rtrans = S.Context.getFunctionNoProtoType(LFT->getReturnType(),
+                                                RFT->getExtInfo());
+    }
+    if (!S.Context.hasSameUnqualifiedType(rtrans, ltrans) &&
+        !S.IsFunctionConversion(rtrans, ltrans))
```
These sorts of boolean conditions always give me a headache, but yes, you’re right, the behaviour would be different if we removed that check here.

brunodf-gf committed:
SemaExpr: further tweaks to checkPointerTypesForAssignment

- Don't transform both ltrans and rtrans when one transform could
suffice.
- Further simplify unneeded code checks that were added to work around
the invalid invocation.
- Fix formatting in test case.

Fznamznon reviewed (commented):
Could you please add a release note?

brunodf-gf committed:
Update for reviewer comments

- Fix typo in comment
- Add release notes entry

brunodf-snps commented:
> Could you please add a release note?

Added.

Fznamznon reviewed (approved):
Thanks!
Let's give @Sirraide a couple of days for additional comments, otherwise LGTM.

Sirraide reviewed (approved):
LGTM, but can you add a few more tests that mix function (pointers) w/ and w/o a prototype where the arguments / return type are compatible, e.g.
```c
void p1(void);
void n1();
void p2(void) __attribute__((noreturn));
void n2() __attribute__((noreturn));

void (*f1)() = p1;
void (*f2)() = n1;
void (*f3)() __attribute__((noreturn)) = p1;
void (*f4)() __attribute__((noreturn)) = n1;

void (*f5)() = p2;
void (*f6)() = n2;
void (*f7)() __attribute__((noreturn)) = p2;
void (*f8)() __attribute__((noreturn)) = n2;
```

brunodf-gf committed:
Add more test cases mixing prototype/no-prototype
